server/util/compression/compression.go

```diff
@@ -161,12 +167,15 @@ func NewZstdDecoderPool() *ZstdDecoderPool {
 	return &ZstdDecoderPool{
 		pool: sync.Pool{
 			New: func() interface{} {
-				dc, err := zstd.NewReader(nil)
+				dc, err := zstd.NewReader(nil, zstd.WithDecoderConcurrency(1))
 				if err != nil {
 					return err
 				}
-				runtime.SetFinalizer(dc, (*zstd.Decoder).Close)
-				return dc
+				ref := &DecoderRef{dc}
+				runtime.SetFinalizer(ref, func(ref *DecoderRef) {
+					ref.Decoder.Close()
+				})
+				return ref
```

Comment:
nit: might be less error prone if you had a NewDecoderRef() function that did this and setup the finalizer internally?

Reply:
oh i guess this only happens in NewZstdDecoderPool, ignore me :)